### PR TITLE
IBP-4901 IBP-5036 save list view

### PR DESCRIPTION
**Pull request by lik-83**

### CODE REVIEW

File: src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java [hunk 1]
```diff
@@ -42,6 +49,15 @@ void addGermplasmEntriesToList(Integer germplasmListId,	SearchCompositeDto<Germp
 
 	void deleteGermplasmListFolder(String cropName, String programUUID, String folderId);
 
+	GermplasmListDto getGermplasmListById(Integer listId);
+
 	List<GermplasmListDto> getGermplasmLists(Integer gid);
 
+	List<GermplasmListSearchResponse> searchGermplasmList(GermplasmListSearchRequest request, Pageable pageable,
+		final String programUUID);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java:57]: remove final

File: src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java [hunk 2]
```diff
@@ -42,6 +49,15 @@ void addGermplasmEntriesToList(Integer germplasmListId,	SearchCompositeDto<Germp
...
+
+	long countSearchGermplasmList(GermplasmListSearchRequest request, final String programUUID);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java:59]: remove final

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListServiceImpl.java [hunk 3]
```diff
@@ -754,6 +748,18 @@ private Integer getFolderIdAsInteger(final String folderId) {
 		return (CROP_LISTS.equals(folderId) || PROGRAM_LISTS.equals(folderId)) ? null : Integer.valueOf(folderId);
 	}
 
+	private static GermplasmListDto transformGermplasmList(GermplasmList input) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListServiceImpl.java:751]: @lik-83 Much better to have a mapper to do this transform

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmListValidator.java [hunk 4]
```diff
@@ -1,33 +1,92 @@
...
-		if (this.errors.hasErrors()) {
+		if (folderName.length() > NAME_MAX_LENGTH) {
+			this.errors.reject("list.folder.name.too.long", new Object[] { NAME_MAX_LENGTH }, "");
 			throw new ApiRequestValidationException(this.errors.getAllErrors());
 		}
 	}
 
+	public void validateListName(final String currentProgram, final String name) {
+		checkNotNull(name, "param.null", new String[] {"name"});
+		checkArgument(name.length() <= 50, "text.field.max.length", new String[] {"name", "50"});
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmListValidator.java:68]: use NAME_MAX_LENGTH

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListDataServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		this.germplasmListValidator.validateGermplasmList(listId);
+
+		return this.germplasmListDataService.getGermplasmListDataTableHeader(listId, programUUID);
+	}
+
+	@Override
+	public void saveGermplasmListDataView(final Integer listId, final List<GermplasmListDataUpdateViewDTO> view) {
+		this.errors = new MapBindingResult(new HashMap<>(), String.class.getName());
+		final GermplasmList germplasmList = this.germplasmListValidator.validateGermplasmList(listId);
+		if (germplasmList.isLockedList()) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListDataServiceImpl.java:65]: In the entries branch,  I created a new function in GermplasmListValidator
```	
     public void validateListIsUnlocked(final GermplasmList germplasmList) {
		if (germplasmList.isLockedList()) {
			this.errors.reject("list.locked", "");
			throw new ApiRequestValidationException(this.errors.getAllErrors());
		}
	}
 ```

You can add this function to the validator and use it here ( I solve the conflicts in the branch), or add a TODO and replace it later

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java [hunk 6]
```diff
@@ -209,4 +228,117 @@ public ResponseEntity deleteGermplasmListFolder(
...
+		@ApiIgnore @PageableDefault(page = PagedResult.DEFAULT_PAGE_NUMBER, size = PagedResult.DEFAULT_PAGE_SIZE) final Pageable pageable) {
+		return new PaginatedSearch().getPagedResult(() -> this.germplasmListDataService.countSearchGermplasmListData(listId, request),
+			() -> this.germplasmListDataService.searchGermplasmListData(listId, request, pageable),
+			pageable);
+	}
+
+	@ApiOperation(value = "Returns a list by a given list id")
+	@RequestMapping(value = "/crops/{cropName}/germplasm-lists/{listId}", method = RequestMethod.GET)
+	@ResponseBody
+	public ResponseEntity<GermplasmListDto> getGermplasmListById(@PathVariable final String cropName,
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:279]: add permissions

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java [hunk 7]
```diff
@@ -209,4 +228,117 @@ public ResponseEntity deleteGermplasmListFolder(
...
+		@PathVariable final Integer listId,
+		@RequestParam(required = false) final String programUUID) {
+		return new ResponseEntity<>(this.germplasmListService.getGermplasmListById(listId), HttpStatus.OK);
+	}
+
+	//TODO: check permission
+	@ApiIgnore
+	@ApiOperation(value = "Locks or unlocks the list depending the current status of it")
+	@RequestMapping(value = "/crops/{cropName}/germplasm-lists/{listId}/toggle-status", method = RequestMethod.POST)
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:288]: add permissions, I will say that we add the same we are using for GET now until Mariano clarifies if we want a specific lock/unlock permission

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java [hunk 8]
```diff
@@ -209,4 +228,117 @@ public ResponseEntity deleteGermplasmListFolder(
...
+		headers.add(HttpHeaders.CONTENT_TYPE, String.format("%s;charset=utf-8", FileUtils.detectMimeType(file.getName())));
+		final FileSystemResource fileSystemResource = new FileSystemResource(file);
+		return new ResponseEntity<>(fileSystemResource, headers, HttpStatus.OK);
+	}
+
+	@ApiIgnore
+	@ApiOperation(value = "Customize the view of the germplasm list data.")
+	@RequestMapping(value = "/crops/{cropName}/germplasm-lists/{listId}/view", method = RequestMethod.PUT)
+	@ResponseBody
+	public ResponseEntity<Void> saveGermplasmListDataView(@PathVariable final String cropName,
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:336]: add permission (same than GET)

**Review by clarysabel — CHANGES_REQUESTED** [review]

File: src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java (hunk 2 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/java/germplasm/GermplasmListService.java]: Done

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListServiceImpl.java (hunk 3 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmListServiceImpl.java]: Done

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java (hunk 6 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:280]: Done

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java (hunk 7 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:288]: Done

File: src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java (hunk 8 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/rest/germplasmlist/GermplasmListResourceGroup.java:334]: Done

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmListValidator.java (hunk 4 above)
Comment (lik-83) [reply src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmListValidator.java]: Done

**Pull request merged**